src/Glimpse.Server.Web/Storage/InMemoryStorage.cs

```diff
@@ -8,65 +7,200 @@ namespace Glimpse.Server.Web
 {
     public class InMemoryStorage : IStorage, IQueryRequests
     {
+        /// <summary>
+        /// Internal class to contain all data associated with a single request. 
+        /// </summary>
+        private class RequestInfo
+        {
+            public RequestInfo(LinkedListNode<Guid> requestIdNode)
+            {
+                this.RequestIdNode = requestIdNode;
+                this.Messages = new LinkedList<IMessage>();
+                this.Indices = null;
+            }
+
+            /// <summary>
+            /// List of individual messages associated with this request
+            /// </summary>
+            public LinkedList<IMessage> Messages { get; private set; }
+            
+            /// <summary>
+            /// The node in the _activeRequests list.  Storing this here allows us to determine to track 
+            /// requests by age in constant time.
+            /// </summary>
+            public LinkedListNode<Guid> RequestIdNode { get; private set; }
+
+            /// <summary>
+            /// Indices associated with this Request
+            /// </summary>
+            public RequestIndices Indices { get; set; }
+        }
+
         public const int RequestsPerPage = 50;
-        public const int MaxRequests = 500; // TODO: Make this configurable
+        public const int DefaultMaxRequests = 500;
 
-        private readonly List<IMessage> _messages;
-        private readonly ConcurrentDictionary<Guid, RequestIndices> _indices;
+        /// <summary>
+        /// Primary storage for Messages. 
+        /// </summary>
+        private readonly Dictionary<Guid, RequestInfo> _requestTable;
 
-        private static readonly object _locker = new object(); // HACK :(
+        /// <summary>
+        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be
+        /// the most recent request to receive a message, the last entry the oldest request to receive a message.
+        /// </summary>
+        private readonly LinkedList<Guid> _activeRequests;
```

Comment:
Sorry for not being clear. I was just wondering what are the chances of someone getting something wrong here if they were making modifications to this class 1/3/6/12 mths from now. Because the two structures need to be kept in sync I was wondering if the exact logic which updates/modifies the objects in `Persist()` could be broken out...

Specially:
```
public void Persist(IMessage message)
{
    var requestId = message.Context.Id;

    lock (_locker)
    {
        var ri = GetRequest(requestId);
        AddMessage(ri, message);

        if (IsRequestMessage(message))
        {
            AddIndices(ri, message);
        }

        TriggerCleanup();
    }
}

private RequestInfo GetRequest(Guid requestId)
{
    RequestInfo ri;
    if (!_requestTable.TryGetValue(requestId, out ri))
    {
        ri = AddRequest(requestId);
    }
    else if (ri.RequestIdNode.Previous != null)
    {
       UpdateRequest(ri);
    }

    return ri;
}

private void RemoveRequest(LinkedListNode<Guid> r)
{
    _activeRequests.Remove(r);
    _requestTable.Remove(r.Value);
}

private RequestInfo AddRequest(Guid requestId)
{
    var llNode = _activeRequests.AddFirst(requestId);

    var ri = new RequestInfo(llNode);
    _requestTable.Add(requestId, ri);

    return ri;
}

private void UpdateRequest(RequestInfo ri)
{
    // move this request to the head of the "active list"
    _activeRequests.Remove(ri.RequestIdNode);
    _activeRequests.AddFirst(ri.RequestIdNode);
}

private void IsRequestMessage(IMessage message)
{
    return message.Context.Type.Equals("request", StringComparison.OrdinalIgnoreCase);
}

private void AddMessage(RequestInfo ri, IMessage message)
{
    // append current message to the message list of this request
    ri.Messages.AddLast(message);
}

private void AddIndices(RequestInfo ri, IMessage message)
{
    if (ri.Indices == null)
    {
        ri.Indices = new RequestIndices(message);
    }
    else
    {
        ri.Indices.Update(message);
    }
}

private void TriggerCleanup()
{
        // clean up old requests
        if (_requestTable.Count > _maxRequests)
        {
            var toRemove = Math.Max(_maxRequests / 10, 1);
            for (int i = 0; i < toRemove; i++)
            {
                RemoveRequest(_activeRequests.Last);
            }
        }
}
```

Note, going through this process I noticed that `TriggerCleanup` should probably happen regardless of if these request `IsRequestMessage`... is that the case?